File: 2021/python/day02/day02.py
import re
from typing import List
from sys import argv


def parse_input(filename: str) -> List[str]:
    return [line.strip() for line in open(filename, "r")]


def part1(instructions: List[str]) -> int:
    position = 0
    depth = 0
    for instruction in instructions:
        value = int(re.search(r"\d+", instruction).group())
        if instruction[0] == "u":
            depth -= value
        elif instruction[0] == "d":
            depth += value
        else:
            position += value
    return position * depth


def part2(instructions: List[str]) -> int:
    position = 0
    depth = 0
    aim = 0
    for instruction in instructions:
        value = int(re.search(r"\d+", instruction).group())
        if instruction[0] == "u":
            aim -= value
        elif instruction[0] == "d":
            aim += value
        else:
            position += value
            depth += aim * value
    return position * depth


def main():
    try:
        input_file = argv[1]
        depths = parse_input(input_file)
        print("PART 1:", part1(depths))
        print("PART 2:", part2(depths))
    except IndexError:
        print("Enter path to data file!")


if __name__ == "__main__":
    main()

File: 2021/go/day02/day02.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

type instruction struct {
	instrType  string
	instrValue int
}

func main() {
	dataFile := flag.String("f", "", "Path to input data file")
	flag.Parse()

	// Process data
	if *dataFile == "" {
		fmt.Fprintln(os.Stderr, "missing input data file")
		os.Exit(1)
	}
	f, err := os.Open(*dataFile)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	scanner := bufio.NewScanner(f)
	instructions, err := scanInputData(scanner)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	// Part 1
	res, err := navigationDepthPosition(instructions)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	fmt.Printf("Part 1: %d\n", res)

	// Part 2
	res, err = navigationDepthPositionAim(instructions)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	fmt.Printf("Part 2: %d\n", res)
}

// scanInputData returns a slice of instruction read from scanner.
func scanInputData(scanner *bufio.Scanner) ([]instruction, error) {
	reType := regexp.MustCompile(`\w+`)
	reValue := regexp.MustCompile(`\d+`)
	var instructions []instruction
	for scanner.Scan() {
		i := scanner.Text()
		instrType := reType.FindString(i)
		instrValue, err := strconv.Atoi(reValue.FindString(i))
		if err != nil {
			return nil, err
		}
		instr := instruction{
			instrType,
			instrValue,
		}
		instructions = append(instructions, instr)
	}
	return instructions, nil
}

func navigationDepthPosition(instructions []instruction) (int, error) {
	depth := 0
	position := 0
	for _, instr := range instructions {
		if instr.instrType == "up" {
			depth -= instr.instrValue
		} else if instr.instrType == "down" {
			depth += instr.instrValue
		} else if instr.instrType == "forward" {
			position += instr.instrValue
		} else {
			return 0, fmt.Errorf("unknown instruction type %s", instr.instrType)
		}
	}
	return depth * position, nil
}

func navigationDepthPositionAim(instructions []instruction) (int, error) {
	depth := 0
	position := 0
	aim := 0
	for _, instr := range instructions {
		if instr.instrType == "up" {
			aim -= instr.instrValue
		} else if instr.instrType == "down" {
			aim += instr.instrValue
		} else if instr.instrType == "forward" {
			position += instr.instrValue
			depth += aim * instr.instrValue
		} else {
			return 0, fmt.Errorf("unknown instruction type %s", instr.instrType)
		}
	}
	return depth * position, nil
}
